File: java/falling.java
import java.util.Scanner;

public class falling {
	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		int input = scan.nextInt();
		int X = -1;
		int minXSum = Integer.MAX_VALUE;
		for (int i = 1; i <= Math.ceil(Math.sqrt(input)); i++) {
			if (input % i == 0) {
				int factor = input / i;
				if ((i + factor) % 2 == 0)
					if (minXSum > i + factor) {
						minXSum = i + factor;
						X = i;
					}
			}
		}
		if (X == -1) {
			System.out.println("impossible");
		} else {
			int Y = input / X;
			int a = (Y + X) / 2;
			int b = (Y - X) / 2;
			System.out.printf("%d %d", b, a);
		}
		scan.close();
	}
}

File: ts/src/falling.ts
import readline = require('readline')

function getFactors (num: number): [number, number] | null {
  for (let i = 1; i <= Math.ceil(Math.sqrt(num)); i++) {
    if (num % i === 0) {
      const factor = num / i
      if ((i + factor) % 2 === 0) {
        return [i, factor]
      }
    }
  }
  return null
}

readline
  .createInterface({
    input: process.stdin,
    output: process.stdout
  })
  .on('line', (line: string) => {
    const num = parseInt(line, 10)
    const factors = getFactors(num)
    if (factors) {
      const [Y, X] = factors
      const n2 = Math.abs(Y + X) / 2
      const n1 = Math.abs(Y - X) / 2
      console.log(`${n1} ${n2}`)
    } else {
      console.log('impossible')
    }
  })
  .on('close', () => {
    process.exit(0)
  })
